plugin/core/windows.py

```diff
@@ -393,15 +393,18 @@ def on_post_exit_async(self, session: Session, exit_code: int, exception: Option
             listener.on_session_shutdown_async(session)
         if exit_code != 0 or exception:
             config = session.config
-            msg = "".join((
-                "{0} exited with status code {1}. ",
-                "Do you want to restart it? If you choose Cancel, it will be disabled for this window for the ",
-                "duration of the current session. ",
-                "Re-enable by running \"LSP: Enable Language Server In Project\" from the Command Palette."
-            )).format(config.name, exit_code)
-            if exception:
-                msg += "\n\n--- Error: ---\n{}".format(str(exception))
-            if sublime.ok_cancel_dialog(msg, "Restart {}".format(config.name)):
+            restart = self._config_manager.record_crash(config.name, exit_code, exception)
+            if not restart:
+                msg = "".join((
+                    "{0} exited with status code {1}. ",
+                    "Do you want to restart it? If you choose Cancel, it will be disabled for this window for the ",
+                    "duration of the current session. ",
+                    "Re-enable by running \"LSP: Enable Language Server In Project\" from the Command Palette."
+                )).format(config.name, exit_code)
+                if exception:
+                    msg += "\n\n--- Error: ---\n{}".format(str(exception))
+                restart = sublime.ok_cancel_dialog(msg, "Restart {}".format(config.name))
```

Comment:
After trying it out I've realized that I would want this text to mention that it crash 5 times during last 3 minutes and also make it a bit more user-friendly. My suggestion would be something like this but we can discuss further:

```suggestion
                msg = "".join((
                    "The {0} server crashed 5 times in the last 3 minutes.\n\n",
                    "You can try to Restart it or you can choose Cancel to disable it for this window for the ",
                    "duration of the current session. ",
                    "Re-enable by running \"LSP: Enable Language Server In Project\" from the Command Palette."
                )).format(config.name)
                if exception:
                    msg += "\n\n--- Error: ---\n{}".format(str(exception))
                restart = sublime.ok_cancel_dialog(msg, "Restart")
```

(Ideally the 5 and 3 variables would be imported of course)

Reply:
Agree!